File: mac/Sources/FineClipboard/Services/SyncCrypto.swift
import Foundation
import CryptoKit
import CArgon2

/// End-to-end encryption for synced clipboard content. The key is derived from the user's
/// sync passphrase with Argon2id and a deterministic salt from the account email, so every
/// device with the same passphrase derives the same key. Byte-for-byte compatible with the
/// Windows client: same Argon2id params (t=3, m=64MiB, p=2) and the same blob layout
/// (nonce(12) || tag(16) || ciphertext) so Mac and Windows decrypt each other's data.
struct SyncCrypto {
    private let key: SymmetricKey

    init(key: Data) { self.key = SymmetricKey(data: key) }

    /// Derives the 256-bit sync key (deterministic across devices).
    static func deriveKey(passphrase: String, email: String) -> Data {
        let salt = Data(SHA256.hash(data: Data("fineclipboard-sync-v1|\(email.trimmingCharacters(in: .whitespaces).lowercased())".utf8)))
        var out = [UInt8](repeating: 0, count: 32)
        let pw = [UInt8](passphrase.utf8)
        let saltBytes = [UInt8](salt)
        let rc = argon2id_hash_raw(3, 65_536, 2, pw, pw.count, saltBytes, saltBytes.count, &out, 32)
        precondition(rc == 0, "argon2id_hash_raw failed: \(rc)")
        return Data(out)
    }

    func encryptToBase64(_ plain: Data) -> String { encrypt(plain).base64EncodedString() }

    func decryptFromBase64(_ b64: String?) -> Data? {
        guard let b64, let blob = Data(base64Encoded: b64) else { return nil }
        return decrypt(blob)
    }

    /// nonce(12) || tag(16) || ciphertext — matches the Windows SyncCrypto layout.
    private func encrypt(_ plain: Data) -> Data {
        let sealed = try! AES.GCM.seal(plain, using: key)
        return Data(sealed.nonce) + sealed.tag + sealed.ciphertext
    }

    private func decrypt(_ blob: Data) -> Data? {
        guard blob.count >= 28 else { return nil }
        let nonceData = blob.prefix(12)
        let tag = blob.subdata(in: 12..<28)
        let ct = blob.subdata(in: 28..<blob.count)
        guard let nonce = try? AES.GCM.Nonce(data: nonceData),
              let box = try? AES.GCM.SealedBox(nonce: nonce, ciphertext: ct, tag: tag) else { return nil }
        return try? AES.GCM.open(box, using: key)
    }
}

File: src/FineClipboard/Services/SyncCrypto.cs
using System;
using System.Security.Cryptography;
using System.Text;
using Konscious.Security.Cryptography;

namespace FineClipboard.Services;

/// <summary>
/// End-to-end encryption for synced clipboard content. The key is derived from the user's
/// sync passphrase with Argon2id, using a deterministic salt from the account email so every
/// device with the same passphrase derives the same key. The key never leaves the device —
/// the server only ever stores AES-256-GCM ciphertext (nonce||tag||ciphertext).
/// </summary>
public sealed class SyncCrypto
{
    private const int NonceLen = 12;
    private const int TagLen = 16;

    private readonly byte[] _key;

    private SyncCrypto(byte[] key) => _key = key;

    public static SyncCrypto FromKey(byte[] key) => new(key);

    /// <summary>Derives the 256-bit sync key from passphrase + email (deterministic across devices).</summary>
    public static byte[] DeriveKey(string passphrase, string email)
    {
        byte[] salt = SHA256.HashData(
            Encoding.UTF8.GetBytes("fineclipboard-sync-v1|" + email.Trim().ToLowerInvariant()));
        using var argon = new Argon2id(Encoding.UTF8.GetBytes(passphrase))
        {
            Salt = salt,
            DegreeOfParallelism = 2,
            MemorySize = 65536,
            Iterations = 3,
        };
        return argon.GetBytes(32);
    }

    public string EncryptToBase64(byte[] plain) => Convert.ToBase64String(Encrypt(plain));

    public byte[]? DecryptFromBase64(string? b64)
    {
        if (string.IsNullOrEmpty(b64))
        {
            return null;
        }
        try
        {
            return Decrypt(Convert.FromBase64String(b64));
        }
        catch
        {
            return null; // wrong passphrase / corrupt data
        }
    }

    private byte[] Encrypt(byte[] plain)
    {
        byte[] nonce = RandomNumberGenerator.GetBytes(NonceLen);
        byte[] cipher = new byte[plain.Length];
        byte[] tag = new byte[TagLen];
        using var aes = new AesGcm(_key, TagLen);
        aes.Encrypt(nonce, plain, cipher, tag);

        byte[] blob = new byte[NonceLen + TagLen + cipher.Length];
        Buffer.BlockCopy(nonce, 0, blob, 0, NonceLen);
        Buffer.BlockCopy(tag, 0, blob, NonceLen, TagLen);
        Buffer.BlockCopy(cipher, 0, blob, NonceLen + TagLen, cipher.Length);
        return blob;
    }

    private byte[] Decrypt(byte[] blob)
    {
        byte[] nonce = new byte[NonceLen];
        byte[] tag = new byte[TagLen];
        Buffer.BlockCopy(blob, 0, nonce, 0, NonceLen);
        Buffer.BlockCopy(blob, NonceLen, tag, 0, TagLen);
        byte[] cipher = new byte[blob.Length - NonceLen - TagLen];
        Buffer.BlockCopy(blob, NonceLen + TagLen, cipher, 0, cipher.Length);

        byte[] plain = new byte[cipher.Length];
        using var aes = new AesGcm(_key, TagLen);
        aes.Decrypt(nonce, cipher, tag, plain);
        return plain;
    }
}
